## E-Commerce Customer Purchase Prediction

#### Life cycle of Machine learning Project

- Understanding the Problem Statement
- Data Collection
- Data Checks to perform
- Exploratory data analysis
- Data Pre-Processing
- Model Training
- Choose best model

### 1) Problem statement
- An e-commerce company wants to predict whether a customer is likely to purchase a product based on their demographic information, income, browsing behavior, previous purchase history, product category, and discount usage.

### 2) Data Collection
- Dataset Source - This data is Genrated by Chatgpt
- The data consists of 10 columns and 50000 rows

### 2.1 Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [5]:
df = pd.read_csv('../data/customer_purchase_merchant_dataset_50000.csv')

In [6]:
df.head(3)

,customer_id,age,gender,income,product_category,browsing_time,pages_viewed,previous_purchases,discount_used,purchased
0,109721,56.0,Male,40702.27,Fashion,28.25,19.0,3.0,Yes,1
1,135306,22.0,Male,14907.71,Grocery,11.25,3.0,4.0,No,1
2,133033,43.0,Female,15860.81,Home,22.55,6.0,6.0,Yes,1


In [7]:
df.shape

(50750, 10)

## Dataset Information
- customer_id: contain numerical values
- age : age of customer
- gender : sex of customer ->(Male/Female)
- income : income of cumtomer
- product_category : customer have many product category --->['Fashion', 'Grocery', 'Home', 'Beauty', nan, 'Electronics',Sports', 'electronics', 'grocery', 'HOME', 'FASHION', 'SPORTS','beauty']
- browsing_time: how much time spend on e-commerce website
- pages_viewed : how much cutomer see the pages
- pervious_purchases : How much customer purchase a project in past
- discount_used : How munch customer take a discount
- purchased : Beased on all customer activate to purchase a product or not --> (0/1)

### 3. Data Checks to perform

- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50750 entries, 0 to 50749
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   customer_id         50750 non-null  int64  
 1   age                 49477 non-null  float64
 2   gender              48924 non-null  object 
 3   income              48471 non-null  float64
 4   product_category    49232 non-null  object 
 5   browsing_time       49483 non-null  float64
 6   pages_viewed        49737 non-null  float64
 7   previous_purchases  48977 non-null  float64
 8   discount_used       49423 non-null  object 
 9   purchased           50750 non-null  int64  
dtypes: float64(5), int64(2), object(3)
memory usage: 3.9+ MB


In [13]:
df.isnull().mean()*100

customer_id           0.000000
age                   2.508374
gender                3.598030
income                4.490640
product_category      2.991133
browsing_time         2.496552
pages_viewed          1.996059
previous_purchases    3.493596
discount_used         2.614778
purchased             0.000000
dtype: float64

### Observation:
The dataset contains null values in 7 columns.

In [14]:
df.nunique()

customer_id           50000
age                      50
gender                    8
income                46892
product_category         12
browsing_time          5522
pages_viewed             41
previous_purchases       13
discount_used             3
purchased                 2
dtype: int64

In [15]:
df.describe()

,customer_id,age,income,browsing_time,pages_viewed,previous_purchases,purchased
count,50750.000000,49477.000000,48471.000000,49483.000000,49737.000000,48977.000000,50750.000000
mean,124998.192906,41.579704,42491.640619,22.010762,8.356555,2.859260,0.530759
std,14436.786782,14.158457,44750.342968,27.750948,11.466022,3.748987,0.499058
min,100001.000000,18.000000,10000.000000,1.000000,1.000000,0.000000,0.000000
25%,112489.250000,29.000000,24026.165000,12.250000,4.000000,2.000000,0.000000
50%,125000.500000,42.000000,34789.320000,18.750000,7.000000,3.000000,1.000000
75%,137496.750000,54.000000,50684.755000,27.070000,10.000000,4.000000,1.000000
max,150000.000000,120.000000,900000.000000,600.000000,250.000000,80.000000,1.000000


In [16]:
## Checking duplicates values
df.duplicated().sum()

np.int64(750)

### Observation:
The dataset contain duplicates values around 750

### Checking inconsisten values for category featues

In [25]:
category_featues = df.select_dtypes(('object'))
for col in category_featues:
    print("Columns:",col)
    print(df[col].unique())
    print()

Columns: gender
['Male' 'Female' 'FEMALE' 'Other' 'male' nan 'other' 'Unknown' 'X']

Columns: product_category
['Fashion' 'Grocery' 'Home' 'Beauty' nan 'Electronics' 'Sports'
 'electronics' 'grocery' 'HOME' 'FASHION' 'SPORTS' 'beauty']

Columns: discount_used
['Yes' 'No' nan 'unknown']



### Observation:-
- Gender contain inconsistent data
- projuct_category cantain a inconsistent data
- discount contain a inconsistent data

### Checking Outliers 

In [29]:
numerical_columns = df.select_dtypes(('int64','float64'))
for col in numerical_columns:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
    
    print(f"{col}: {len(outliers)} outliers")

customer_id: 0 outliers
age: 60 outliers
income: 2190 outliers
browsing_time: 1313 outliers
pages_viewed: 1262 outliers
previous_purchases: 445 outliers
purchased: 0 outliers


### Observation:
- The Dataset contain outliers in 4 columns and the columns name is:
- age: 60 outliers
- income: 2190 outliers
- browsing_time: 1313 outliers
- pages_viewed: 1262 outliers
- previous_purchases: 445 outliers

### Conclusion

During the initial data analysis, the following data quality issues were identified:

- Outliers are present in the dataset.
- Missing values are present in several columns.
- Inconsistent values were identified in some columns.
- Data types of all columns were checked.
- A basic understanding of the dataset was established.